From 07c3e56e2d4cc4348c559f4998ffba85fb0cf600 Mon Sep 17 00:00:00 2001
From: Alexey Minnekhanov <alexeymin@postmarketos.org>
Date: Tue, 28 May 2024 07:24:00 +0300
Subject: [PATCH 02/20] HACK: of: fdt: filter out dangerous Android bootloader
 parameters

Since Android primary bootloader with system-as-root feature [1]
enabled adds some arguments to kernel command line from boot.img,
that make booting Linux harder:

  * skip_initramfs
  * root=/dev/dm-0
  * init=/init
  * dm=...

Most importantly, Linux processes root= parameter and at boot time
skips initramfs and tries to mount root device that does not exist
and fails.

Add a quick hack to scramble these parameters from cmdline passed by
primary bootloader in FDT /chosen/bootargs property, so that Linux
won't accidentally process them.

[1] https://source.android.com/docs/core/architecture/partitions/system-as-root

Signed-off-by: Alexey Minnekhanov <alexeymin@postmarketos.org>
---
 arch/arm64/Kconfig | 26 ++++++++++++++++++++++++++
 drivers/of/fdt.c   | 16 ++++++++++++++++
 2 files changed, 42 insertions(+)

diff --git a/arch/arm64/Kconfig b/arch/arm64/Kconfig
index fc57d90d9..f68b5c3cd 100644
--- a/arch/arm64/Kconfig
+++ b/arch/arm64/Kconfig
@@ -2465,6 +2465,32 @@ config CMDLINE_FORCE
 
 endchoice
 
+config CMDLINE_DROP_DANGEROUS_ANDROID_OPTIONS
+	bool "Drop certain dangerous options from cmdline"
+	default n
+	help
+	  Android >=9 primary bootloader with system-as-root feature [1]
+	  enabled passes some arguments in kernel command line, that make
+	  booting Linux harder:
+
+	    * skip_initramfs
+	    * root=/dev/dm-0
+	    * init=/init
+	    * dm=...
+
+	  Those parameters override default boot partition to hardcoded,
+	  set init binary to /init, disable booting from initramfs.
+
+	  Most importantly, Linux processes root= parameter and at boot time
+	  skips initramfs and tries to mount device that does not exist and
+	  fails.
+
+	  If this option is enabled, those cmdline parameters will be erased
+	  from bootloader's command line, and custom OS can boot the way it
+	  likes.
+
+	  [1] https://source.android.com/docs/core/architecture/partitions/system-as-root
+
 config EFI_STUB
 	bool
 
diff --git a/drivers/of/fdt.c b/drivers/of/fdt.c
index 26f66046c..335321ab9 100644
--- a/drivers/of/fdt.c
+++ b/drivers/of/fdt.c
@@ -1093,6 +1093,9 @@ int __init early_init_dt_scan_chosen(char *cmdline)
 	const char *p;
 	const void *rng_seed;
 	const void *fdt = initial_boot_params;
+#ifdef CONFIG_CMDLINE_DROP_DANGEROUS_ANDROID_OPTIONS
+	char *cmdline2;
+#endif
 
 	node = fdt_path_offset(fdt, "/chosen");
 	if (node < 0)
@@ -1143,6 +1146,19 @@ int __init early_init_dt_scan_chosen(char *cmdline)
 #endif
 #endif /* CONFIG_CMDLINE */
 
+#ifdef CONFIG_CMDLINE_DROP_DANGEROUS_ANDROID_OPTIONS
+	/* Quick hack - replace first character with '_' */
+	cmdline2 = strstr((const char *)cmdline, "skip_initramfs");
+	if (cmdline2)
+		*cmdline2 = '_';
+	cmdline2 = strstr((const char *)cmdline, "root=");
+	if (cmdline2)
+		*cmdline2 = '_';
+	cmdline2 = strstr((const char *)cmdline, "init=");
+	if (cmdline2)
+		*cmdline2 = '_';
+#endif
+
 	pr_debug("Command line is: %s\n", (char *)cmdline);
 
 	return 0;
-- 
2.56.0

